File: duck_vlm/state.py
"""MuJoCo state adapter for VLM prompts and reward/recording data."""
from __future__ import annotations

import math
from typing import Any

import numpy as np

from .scenes import SceneSpec
from .types import DuckState


def _quat_rotate_inverse(quat: np.ndarray, vec: np.ndarray) -> np.ndarray:
    w = float(quat[0])
    xyz = np.asarray(quat[1:4], dtype=np.float64)
    t = np.cross(xyz, vec) * 2.0
    return vec - w * t + np.cross(xyz, t)


_TARGET_ALIASES = {
    "ball": "ball", "sphere": "ball", "orange ball": "ball", "足球": "ball", "球": "ball",
    "beacon": "beacon", "marker": "beacon", "信标": "beacon", "目标": "beacon",
    "owner": "owner", "person": "owner", "主人": "owner", "人": "owner",
}


class DuckStateSensor:
    def __init__(self, sim: Any, scene: SceneSpec | None = None):
        self.sim = sim
        self.scene = scene
        self._target_cache: dict[str, int | None] = {}

    def resolve_target(self, target_name: str) -> int | None:
        key = (target_name or "ball").strip().lower()
        canonical = _TARGET_ALIASES.get(key, key)
        if self.scene is not None:
            resolved = self.scene.resolve_body(target_name)
            if resolved:
                canonical = resolved
        cache_key = f"{target_name}|{canonical}"
        if cache_key in self._target_cache:
            return self._target_cache[cache_key]
        body_id: int | None = None
        if canonical == "ball":
            body_id = int(getattr(self.sim, "ball_body", -1))
        if canonical == "owner":
            for candidate in ("owner", "person", "human"):
                try:
                    import mujoco
                    body_id = int(mujoco.mj_name2id(self.sim.model, mujoco.mjtObj.mjOBJ_BODY, candidate))
                except Exception:
                    body_id = -1
                if body_id is not None and body_id >= 0:
                    break
        if body_id is None or body_id < 0:
            try:
                import mujoco
                body_id = int(mujoco.mj_name2id(self.sim.model, mujoco.mjtObj.mjOBJ_BODY, canonical))
            except Exception:
                body_id = -1
        if body_id is not None and body_id < 0:
            body_id = None
        self._target_cache[cache_key] = body_id
        return body_id

    def snapshot(self, target_name: str = "ball", sim_time: float = 0.0) -> DuckState:
        d = self.sim.data
        trunk_id = int(self.sim.trunk_id)
        trunk = np.asarray(d.xpos[trunk_id], dtype=np.float64).copy()
        quat = np.asarray(d.xquat[trunk_id], dtype=np.float64).copy()
        xmat = np.asarray(d.xmat[trunk_id], dtype=np.float64).reshape(3, 3)
        heading = math.atan2(float(xmat[1, 0]), float(xmat[0, 0]))
        gravity = _quat_rotate_inverse(quat, np.array([0.0, 0.0, -1.0]))
        speed = float(np.linalg.norm(np.asarray(d.qvel[:2], dtype=np.float64)))
        gyro_adr = int(getattr(self.sim, "imu_gyro_adr", 0))
        gyro = np.asarray(d.sensordata[gyro_adr:gyro_adr + 3], dtype=np.float64)
        state = DuckState(
            sim_time=float(sim_time), x=float(trunk[0]), y=float(trunk[1]), z=float(trunk[2]),
            heading_rad=float(heading), linear_speed_mps=speed,
            angular_speed_rps=float(np.linalg.norm(gyro)), upright=float(-gravity[2]),
            fallen=bool((-gravity[2]) < 0.5), target_name=target_name or "ball",
        )
        target_id = self.resolve_target(target_name)
        if target_id is None:
            return state
        target_world = np.asarray(d.xpos[target_id], dtype=np.float64).copy()
        target_local = _quat_rotate_inverse(quat, target_world - trunk)
        in_front = bool(target_local[0] > 0.03)
        uv = self._project_to_headcam(target_world) if in_front else None
        if uv is not None and self._is_occluded(target_world, target_id):
            uv = None
        state.target_uv = uv
        state.target_visible = uv is not None
        if state.target_visible:
            # The head camera has the target, so a detector + depth sensor could
            # report the same range/bearing. This is the only case where the
            # privileged simulator state is allowed into the prompt.
            state.target_world_xyz = tuple(float(x) for x in target_world[:3])
            state.target_range_m = float(np.linalg.norm(target_local))
            state.target_bearing_rad = float(math.atan2(target_local[1], target_local[0]))
            state.target_elevation_rad = float(math.atan2(target_local[2], math.hypot(target_local[0], target_local[1])))
        else:
            # Occluded or out of frame: there is no depth/range sensor on the duck,
            # so withhold ground truth instead of leaking the answer.
            state.target_world_xyz = None
            state.target_range_m = None
            state.target_bearing_rad = None
            state.target_elevation_rad = None
        return state
    def _is_occluded(self, target_world: np.ndarray, target_id: int) -> bool:
        """True when scene geometry blocks the head camera's line of sight."""
        try:
            import mujoco
            from duck_play.perception.camera import DuckHeadCam
            cam = DuckHeadCam(width=320, height=240, pitch_up_deg=-20, fwd_m=0.09, up_m=0.03)
            origin = np.asarray(cam.pose(self.sim)[0], dtype=np.float64)
        except Exception:
            return False
        direction = np.asarray(target_world, dtype=np.float64) - origin
        dist = float(np.linalg.norm(direction))
        if dist < 1e-6:
            return False
        direction = direction / dist
        geomid = np.array([-1], dtype=np.int32)
        try:
            hit = float(mujoco.mj_ray(self.sim.model, self.sim.data, origin, direction,
                                      None, 1, -1, geomid))
        except Exception:
            return False
        if hit < 0.0 or hit >= dist - 0.02:
            return False
        if hit < 0.05:
            return False  # the duck's own head/neck shell
        gid = int(geomid[0])
        if gid < 0:
            return False
        return int(self.sim.model.geom_bodyid[gid]) != int(target_id)

    def _project_to_headcam(self, world_xyz: np.ndarray) -> tuple[float, float] | None:
        try:
            from duck_play.perception.camera import DuckHeadCam
            return DuckHeadCam(width=320, height=240, pitch_up_deg=-20, fwd_m=0.09, up_m=0.03).project(world_xyz, self.sim)
        except Exception:
            return None

def reset_to_spawn(sim: Any, task: dict | None = None, rng: Any = None) -> dict:
    """Reset the sim and place the duck at the task's declared ``robot_spawn``.

    Consecutive episodes otherwise inherit whatever pose the previous task ended
    in, which silently makes some tasks trivial (e.g. walk_turn_stop passing in
    three steps because the duck already stood near the goal region). The scene
    pack declares a spawn per task, so honour it.
    """
    import random as _random
    rand = rng or _random
    if hasattr(sim, "reset"):
        sim.reset()
    info: dict[str, Any] = {"reset": True, "spawn": None}
    spawn = (task or {}).get("robot_spawn") or {}
    xy = spawn.get("xy")
    if not (isinstance(xy, (list, tuple)) and len(xy) >= 2):
        return info
    try:
        import mujoco
    except Exception:
        return info
    x, y = float(xy[0]), float(xy[1])
    radius = float(spawn.get("random_xy_radius_m") or 0.0)
    if radius > 0.0:
        ang = rand.uniform(0.0, 2.0 * math.pi)
        r = radius * math.sqrt(rand.random())
        x += r * math.cos(ang)
        y += r * math.sin(ang)
    yaw = math.radians(float(spawn.get("yaw_deg") or 0.0))
    jitter = float(spawn.get("random_yaw_deg") or 0.0)
    if jitter > 0.0:
        yaw += math.radians(rand.uniform(-jitter, jitter))
    z = float(spawn.get("z_m") or 0.125)
    adr = getattr(sim, "qpos_adr", None)
    if adr is None:
        return info
    sim.data.qpos[adr:adr + 7] = [x, y, z, math.cos(yaw / 2.0), 0.0, 0.0, math.sin(yaw / 2.0)]
    sim.data.qvel[:] = 0.0
    mujoco.mj_forward(sim.model, sim.data)
    info["spawn"] = {"xy": [round(x, 3), round(y, 3)], "yaw_deg": round(math.degrees(yaw), 1)}
    return info

File: web_sim/src/state.js
/**
 * 本体感知 + 目标可达性 —— `duck_vlm/state.py` + `duck_vlm/scenes.py` 的浏览器版。
 *
 * 最重要的一条设计（用户明确要求过）：**鸭子没看到目标时，不许用上帝视角真值。**
 * 只有目标真的落在头摄画面里、并且视线上没有遮挡，才把距离/方位写进提示词；
 * 否则那些字段是 `not_found`，模型只能靠转头扫视去找。
 * 鸭子身上没有深度传感器，这条规则就是“仿真真值不许偷偷泄漏给 VLM”的闸门。
 */

function quatRotateInverse(quat, vec) {
  const w = quat[0], x = quat[1], y = quat[2], z = quat[3];
  const t = [2 * (y * vec[2] - z * vec[1]), 2 * (z * vec[0] - x * vec[2]), 2 * (x * vec[1] - y * vec[0])];
  return [vec[0] - w * t[0] + (y * t[2] - z * t[1]),
          vec[1] - w * t[1] + (z * t[0] - x * t[2]),
          vec[2] - w * t[2] + (x * t[1] - y * t[0])];
}

export class DuckStateSensor {
  /**
   * @param {object} opts
   * @param {object} opts.duck DuckSim
   * @param {object|null} opts.scene SceneIndex
   * @param {object|null} opts.view DuckView —— 给了就用**渲染器**判断目标有没有被挡住
   *        （这个 WASM 构建的 mj_ray 拿不到 geomid，见 view.maskVisible 的注释）
   */
  constructor({ duck, scene = null, view = null }) {
    this.duck = duck;
    this.scene = scene;
    this.view = view;
    this._cache = new Map();
    this.mujoco = duck.mujoco;
    this.model = duck.model;
    this.data = duck.data;
    this._BODY = duck.mujoco.mjtObj.mjOBJ_BODY.value;
  }

  /** 目标名 → body id。先问场景元数据（能理解自然语言标签），再直接当 body 名查。 */
  resolveTarget(targetName) {
    const raw = String(targetName || "ball").trim();
    const key = raw.toLowerCase();
    if (this._cache.has(key)) return this._cache.get(key);
    let body = this.scene ? this.scene.resolveBody(raw) : null;
    if (!body && key === "ball") body = "ball";
    let id = -1;
    if (body) id = this.mujoco.mj_name2id(this.model, this._BODY, body);
    if (id < 0) id = this.mujoco.mj_name2id(this.model, this._BODY, key);
    const out = id < 0 ? null : id;
    this._cache.set(key, out);
    return out;
  }

  /**
   * 采一帧本体状态。sim_time 由调用方给（单位秒）。
   * @returns {object} DuckState（字段名与 Python types.DuckState 一致）
   */
  snapshot(targetName = "ball", simTime = 0) {
    const d = this.data, t = this.duck.bodies.trunk;
    const trunk = [d.xpos[t * 3], d.xpos[t * 3 + 1], d.xpos[t * 3 + 2]];
    const quat = d.xquat.subarray(t * 4, t * 4 + 4);
    const m = d.xmat;
    const heading = Math.atan2(m[t * 9 + 3], m[t * 9 + 0]);
    const gravity = quatRotateInverse(quat, [0, 0, -1]);
    const upright = -gravity[2];
    const gyro = this.duck.gyroAdr;
    const state = {
      simTime: Number(simTime), x: trunk[0], y: trunk[1], z: trunk[2],
      headingRad: heading,
      linearSpeedMps: Math.hypot(d.qvel[0], d.qvel[1]),
      angularSpeedRps: Math.hypot(d.sensordata[gyro], d.sensordata[gyro + 1], d.sensordata[gyro + 2]),
      upright, fallen: upright < 0.5,
      targetName: targetName || "ball",
      targetVisible: false, targetRangeM: null, targetBearingRad: null,
      targetElevationRad: null, targetUv: null, targetWorldXyz: null,
    };

    const id = this.resolveTarget(targetName);
    if (id === null) return state;
    const world = [d.xpos[id * 3], d.xpos[id * 3 + 1], d.xpos[id * 3 + 2]];
    const local = quatRotateInverse(quat, [world[0] - trunk[0], world[1] - trunk[1], world[2] - trunk[2]]);
    const inFront = local[0] > 0.03;
    let uv = inFront ? this.duck.projectToHeadCam(world) : null;
    if (uv) {
      if (this.view) {
        // 画面里一个像素都没有 = 真的看不见（出画或被挡住）
        const mask = this.view.maskVisible(this.duck.geomIndicesUnderBody(id));
        if (mask.count < 2) uv = null;
      } else if (this.duck.isOccluded(world, id)) {
        uv = null;
      }
    }
    state.targetUv = uv ? [uv.u, uv.v] : null;
    state.targetVisible = uv !== null;
    if (state.targetVisible) {
      // 只有在头摄真的拍到目标时才允许把仿真真值写进提示词。
      state.targetWorldXyz = world;
      state.targetRangeM = Math.hypot(local[0], local[1], local[2]);
      state.targetBearingRad = Math.atan2(local[1], local[0]);
      state.targetElevationRad = Math.atan2(local[2], Math.hypot(local[0], local[1]));
    }
    return state;
  }

  /** 提示词里的 ROBOT STATE 一行（格式与 Python DuckState.prompt_text 完全一致）。 */
  static promptText(s) {
    const parts = [
      `body=(${s.x.toFixed(2)},${s.y.toFixed(2)},${s.z.toFixed(2)})`,
      `heading=${s.headingRad.toFixed(2)}rad`,
      `speed=${s.linearSpeedMps.toFixed(2)}m/s`,
      `turn_rate=${s.angularSpeedRps.toFixed(2)}rad/s`,
      `upright=${s.upright.toFixed(2)}`,
      "fallen=" + (s.fallen ? "yes" : "no"),
    ];
    if (s.targetRangeM === null) {
      parts.push(`target(${s.targetName})=not_found`);
    } else {
      parts.push(`target(${s.targetName})=range ${s.targetRangeM.toFixed(2)}m, ` +
                 `bearing ${s.targetBearingRad.toFixed(2)}rad, ` +
                 `visible=${s.targetVisible ? "yes" : "no"}`);
    }
    return parts.join(" | ");
  }
}
